lang/rust/avro/src/types.rs

```diff
@@ -2391,4 +2442,150 @@ Field with name '"b"' is not a member of the map items"#,
             "field b record is invalid against the schema"
         ); // this should pass, but doesn't
     }
+
+    #[test]
+    fn test_avro_3674_validate_namespace_resolution() {
+        use crate::ser::Serializer;
+        use serde::Serialize;
+
+        let schema = Schema::parse_str(
+            r#"{
+            "type": "record",
+            "name": "NamespacedMessage",
+            "namespace": "com.domain",
+            "fields": [
+                {
+                    "type": "record",
+                    "name": "field_a",
+                    "fields": [
+                        {
+                            "name": "enum_a",
+                            "type": {
+                                "type": "enum",
+                                "name": "EnumType",
+                                "symbols": [
+                                    "SYMBOL_1",
+                                    "SYMBOL_2"
+                                ],
+                                "default": "SYMBOL_1"
+                            }
+                        },
+                        {
+                            "name": "enum_b",
+                            "type": "EnumType"
+                        }
+                    ]
+                }
+            ]
+        }"#,
+        )
+        .unwrap();
+
+        #[derive(Serialize)]
+        enum EnumType {
+            #[serde(rename = "SYMBOL_1")]
+            Symbol1,
+            #[serde(rename = "SYMBOL_2")]
+            Symbol2,
+        }
+
+        #[derive(Serialize)]
+        struct FieldA {
+            enum_a: EnumType,
+            enum_b: EnumType,
+        }
+
+        #[derive(Serialize)]
+        struct NamespacedMessage {
+            field_a: FieldA,
+        }
+
+        let msg = NamespacedMessage {
+            field_a: FieldA {
+                enum_a: EnumType::Symbol2,
+                enum_b: EnumType::Symbol1,
+            },
+        };
+
+        let mut ser = Serializer::default();
+        let test_value: Value = msg.serialize(&mut ser).unwrap();
+        assert!(test_value.validate(&schema), "test_value should validate");
+        assert!(
+            test_value.resolve(&schema).is_ok(),
+            "test_value should resolve"
+        );
+    }
+
+    #[test]
+    fn test_avro_3674_validate_no_namespace_resolution() {
```

Comment:
@martin-g we could get rid of some code duplication by unifying this test-case and the one above as the only difference between the two should be the `namespace` in the schema.

Reply:
I may do it in a separate PR!
I want to do some cleanup to the `_internal` methods too.